text/0000-convert-stringify-to-async.md

```diff
@@ -0,0 +1,67 @@
+*   Start date: 2022-03-13
+*   Scope: unifiedjs/unified-engine
+*   RFC PR: <!-- leave this empty -->
+*   Implementation issue(s): <!-- leave this empty -->
+
+# Summary
+
+Update the `stringify` function to be an async function.
+
+## Motivation
+
+Why are we doing this?
+The change allows for great fexibliy in implementing Compile plugins, by allowing `stringify`
+to return a promise. Additionally, this will ensure that plugins loaded via cli (unified-args)
+function the same as plugins loaded via `.use()`.  
+What use cases does it support?  
+Cases when Compile is naturally async.  
```

Comment:
You are precluding the use of stream.Writable , which is a common why to build up a string.  Also prevents the use of a micro-service, or as any lib that might need a resource from the internet. 

You stated somewhere else that pdfMake has a sync functionality, if it does I can not find it; give that pdfMake is based on pdfKit which is based on node streams.

Reply:
> You are precluding the use of stream.Writable

Yes. Because streaming and ASTs don‘t make sense. PostCSS doesn’t stream. Babel doesn’t stream.

Streaming makes sense because some chunk of data (e.g., binary, text) on disk is (or can be) *soooo* big that it would become slow to handle in one go and better to handle it in chunks.
But, async work is always slower. `fs.readFileSync()` is always faster than `fs.readFile`. But the second has other benefits.

unified has access to the whole tree (the AST) already *in memory*. The size of the AST in memory is much bigger than the source of the document. Because we have **everything** already, serializing that tree to plain-text **never has to be sync**.

> which is a common why to build up a string.

Streams and ASTs are not commonly combined.

>  Also prevents the use of a micro-service, or as any lib that might need a resource from the internet.

No. Not at all. **Streaming exists outside of unified**. Note that you’re talking about **parsing here**, where this issue/RFC is about serializing.